dpnp/dpnp_iface_linearalgebra.py

```diff
@@ -269,63 +285,364 @@ def matmul(x1, x2, out=None, **kwargs):
     >>> np.matmul(a, b)
     array([[4, 1],
            [2, 2]])
 
+    For 2-D mixed with 1-D, the result is the usual.
+
+    >>> a = np.array([[1, 0], [0, 1]])
+    >>> b = np.array([1, 2])
+    >>> np.matmul(a, b)
+    array([1, 2])
+    >>> np.matmul(b, a)
+    array([1, 2])
+
+    Broadcasting is conventional for stacks of arrays
+
+    >>> a = np.arange(2 * 2 * 4).reshape((2, 2, 4))
+    >>> b = np.arange(2 * 2 * 4).reshape((2, 4, 2))
+    >>> np.matmul(a,b).shape
+    (2, 2, 2)
+    >>> np.matmul(a, b)[0, 1, 1]
+    array(98)
+    >>> np.sum(a[0, 1, :] * b[0 , :, 1])
+    array(98)
+
+    Vector, vector returns the scalar inner product, but neither argument is complex-conjugated:
+
+    >>> x1 = np.array([2j, 3j])
+    >>> x2 = np.array([2j, 3j])
+    >>> np.matmul(x1, x2)
+    array(-13+0j)
+
+    The ``@`` operator can be used as a shorthand for ``matmul`` on
+    :class:`dpnp.ndarray`.
+
+    >>> x1 @ x2
+    array(-13+0j)
+
     """
 
-    x1_desc = dpnp.get_dpnp_descriptor(x1, copy_when_nondefault_queue=False)
-    x2_desc = dpnp.get_dpnp_descriptor(x2, copy_when_nondefault_queue=False)
-    if x1_desc and x2_desc and not kwargs:
-        if x1_desc.ndim != 2 or x2_desc.ndim != 2:
-            pass
-        elif not x1_desc.ndim:
-            pass
-        elif not x2_desc.ndim:
-            pass
-        elif not x1_desc.size:
-            pass
-        elif not x2_desc.size:
+    dpnp.check_supported_arrays_type(x1)
+    dpnp.check_supported_arrays_type(x2)
+
+    if subok is False:
+        raise NotImplementedError(
+            "subok keyword argument is only supported by its default value."
+        )
+    else:
+        x1_ndim = x1.ndim
+        x2_ndim = x2.ndim
+
+        if x1_ndim == 0 or x2_ndim == 0:
+            raise ValueError(
+                "matmul: Input operand does not have enough dimensions"
+            )
+
+        exec_q = dpctl.utils.get_execution_queue((x1.sycl_queue, x2.sycl_queue))
+        if exec_q is None:
+            raise ValueError(
+                "Execution placement can not be unambiguously inferred "
+                "from input arguments."
+            )
+
+        # input arrays should be C_CONTIGUOUS or F_CONTIGUOUS
+        if not x1.flags["C_CONTIGUOUS"] and not x1.flags["F_CONTIGUOUS"]:
+            v = dpnp.empty_like(x1, order="C")
+            ht_copy_ev_x1, _ = ti._copy_usm_ndarray_into_usm_ndarray(
+                src=dpnp.get_usm_ndarray(x1),
+                dst=v.get_array(),
+                sycl_queue=x1.sycl_queue,
+            )
+            x1 = v
+            ht_copy_ev_x1.wait()
+
+        if not x2.flags["C_CONTIGUOUS"] and not x2.flags["F_CONTIGUOUS"]:
+            v = dpnp.empty_like(x2, order="C")
+            ht_copy_ev_x2, _ = ti._copy_usm_ndarray_into_usm_ndarray(
+                src=dpnp.get_usm_ndarray(x2),
+                dst=v.get_array(),
+                sycl_queue=x2.sycl_queue,
+            )
+            x2 = v
+            ht_copy_ev_x2.wait()
```

Comment:
[Tag: copy]
You do not have to do copying at the beginning of  `matmul` function
It would be better to do it before calculating the result
```suggestion
```